activerecord/lib/active_record/connection_adapters/persisted_schema_cache.rb

```diff
@@ -0,0 +1,118 @@
+module ActiveRecord
+  module ConnectionAdapters
+    class PersistedSchemaCache
+      attr_accessor :connection
+      attr_reader :version
+
+      class ImmutableError < StandardError
+        def initialize
+          super("PersistedSchemaCache cannot be mutated")
+        end
+      end
+
+      def initialize(connection, version:, columns:, columns_hash:, primary_keys:)
+        @version      = version
+        @columns      = columns
+        @columns_hash = columns_hash
+        @primary_keys = primary_keys
+        @data_sources = Set.new(columns.keys)
+      end
+
+      def initialize_dup(other)
+        super
+        @columns      = @columns.dup
+        @columns_hash = @columns_hash.dup
+        @primary_keys = @primary_keys.dup
+        @data_sources = @data_sources.dup
+      end
+
+      def primary_keys(table_name)
+        @primary_keys[table_name]
+      end
+
+      # A cached lookup for table existence.
+      def data_source_exists?(name)
+        @data_sources.include?(name)
+      end
+      alias table_exists? data_source_exists?
+      deprecate table_exists?: "use #data_source_exists? instead"
+
+      # Add internal cache for table with +table_name+.
+      def add(table_name)
+        raise ImmutableError
+      end
+
+      def data_sources(name)
+        @data_sources.include?(name)
+      end
+      alias tables data_sources
+      deprecate tables: "use #data_sources instead"
+
+      # Get the columns for a table
+      def columns(table_name)
+        @columns[table_name]
+      end
+
+      # Get the columns for a table as a hash, key is the column name
+      # value is the column object.
+      def columns_hash(table_name)
+        @columns_hash[table_name]
+      end
+
+      # Clears out internal caches
+      def clear!
```

Comment:
Because it's a public method that someone may call, and since this class is immutable, we'd like to avoid that call.